fraud/service.go

```diff
@@ -33,24 +32,29 @@ func NewService(p *pubsub.PubSub, getter headerFetcher) Service {
 
 func (f *service) Subscribe(proofType ProofType) (Subscription, error) {
 	// TODO: @vgonkivs check if fraud proof is in fraud store, then return with error
-	f.mu.RLock()
+	f.mu.Lock()
 	u, ok := f.unmarshalers[proofType]
-	f.mu.RUnlock()
 	if !ok {
 		return nil, errors.New("fraud: unmarshaler is not registered")
 	}
 
-	t, wasNotjoined, err := f.topics.getTopic(proofType)
-	if err != nil {
-		return nil, err
-	}
-	// if topic was joined for the first time then we should register a validator for it.
-	if wasNotjoined {
-		if err = f.topics.registerValidator(proofType, f.processIncoming); err != nil {
+	t, ok := f.topics.getTopic(proofType)
+
+	// if topic was not stored in cache, then we should register a validator and
+	// join pubsub topic.
+	if !ok {
+		err := f.topics.registerValidator(proofType, f.processIncoming)
+		if err != nil {
+			f.mu.Unlock()
+			return nil, err
+		}
+		t, err = f.topics.join(proofType)
+		if err != nil {
+			f.mu.Unlock()
 			return nil, err
 		}
 	}
-
+	f.mu.Unlock()
```

Comment:
Since this is at the end anyway, you can just defer `f.mu.Unlock` after line 35 and remove the unlocks at line 48 and line 53

Reply:
In this case unlock will be called after `newSubscription`. But `newSubscription` subscribes to the given topic and it can take sometime. I think it's better to manually unlock mutex before calling `newSubscription`